File: voice_input/stats.py
import json
from dataclasses import dataclass
from datetime import datetime, timedelta
from pathlib import Path


@dataclass
class UsageStats:
    session_count: int = 0
    total_seconds: float = 0.0
    total_chars: int = 0

    @property
    def total_minutes_int(self) -> int:
        return int(self.total_seconds // 60)

    @property
    def avg_chars_per_minute(self) -> float:
        if self.total_seconds <= 0:
            return 0.0
        return self.total_chars / (self.total_seconds / 60)


def count_text_chars(text: str) -> int:
    # 统计时忽略空白，更接近“实际输入了多少文字”
    return len("".join(str(text).split()))


class StatsStore:
    def __init__(self, path: str | Path) -> None:
        self.path = Path(path)

    def append_event(self, *, text: str, duration_seconds: float, created_at: datetime | None = None) -> None:
        event = {
            "created_at": (created_at or datetime.now()).isoformat(timespec="seconds"),
            "duration_seconds": round(max(0.0, float(duration_seconds)), 2),
            "text_chars": count_text_chars(text),
        }
        self.path.parent.mkdir(parents=True, exist_ok=True)
        with self.path.open("a", encoding="utf-8") as file:
            file.write(json.dumps(event, ensure_ascii=False) + "\n")

    def summarize_since(self, start_time: datetime) -> UsageStats:
        stats = UsageStats()
        for event in self._iter_events():
            created_at = event.get("created_at")
            if created_at is None or created_at < start_time:
                continue
            stats.session_count += 1
            stats.total_seconds += float(event.get("duration_seconds", 0.0) or 0.0)
            stats.total_chars += int(event.get("text_chars", 0) or 0)
        return stats

    def summarize_recent_days(self, days: int) -> UsageStats:
        return self.summarize_since(datetime.now() - timedelta(days=days))

    def summarize_recent_hours(self, hours: int) -> UsageStats:
        return self.summarize_since(datetime.now() - timedelta(hours=hours))

    def summarize_by_day(self, days: int = 7) -> list[tuple[str, UsageStats]]:
        today = datetime.now().date()
        stats_by_day = {
            (today - timedelta(days=offset)).isoformat(): UsageStats()
            for offset in range(days)
        }
        for event in self._iter_events():
            created_at = event.get("created_at")
            if created_at is None:
                continue
            day_key = created_at.date().isoformat()
            day_stats = stats_by_day.get(day_key)
            if day_stats is None:
                continue
            day_stats.session_count += 1
            day_stats.total_seconds += float(event.get("duration_seconds", 0.0) or 0.0)
            day_stats.total_chars += int(event.get("text_chars", 0) or 0)
        return sorted(stats_by_day.items())

    def _iter_events(self):
        if not self.path.exists():
            return
        with self.path.open("r", encoding="utf-8") as file:
            for line in file:
                line = line.strip()
                if not line:
                    continue
                try:
                    raw = json.loads(line)
                    created_at = datetime.fromisoformat(str(raw["created_at"]))
                    yield {
                        "created_at": created_at,
                        "duration_seconds": raw.get("duration_seconds", 0.0),
                        "text_chars": raw.get("text_chars", 0),
                    }
                except Exception:
                    continue

File: src-tauri/src/stats.rs
use chrono::{DateTime, Duration, Local};
use serde::{Deserialize, Serialize};
use serde_json::json;
use std::io::Write;
use std::path::PathBuf;

#[derive(Debug, Clone, Serialize)]
pub struct UsageStats {
    pub session_count: u32,
    pub total_seconds: f64,
    pub total_chars: u32,
    pub total_minutes_int: u32,
    pub avg_chars_per_minute: f64,
}

#[derive(Debug, Clone, Serialize)]
pub struct DailyUsageStats {
    pub day: String,
    pub stats: UsageStats,
}

#[derive(Debug, Clone, Serialize)]
pub struct StatsSnapshot {
    pub path: String,
    pub recent_24h: UsageStats,
    pub recent_7d: UsageStats,
    pub by_day: Vec<DailyUsageStats>,
    pub history: Vec<HistoryEvent>,
}

#[derive(Debug, Clone, Serialize)]
pub struct HistoryEvent {
    pub created_at: String,
    pub duration_seconds: f64,
    pub text_chars: u32,
}

#[derive(Debug, Deserialize)]
struct RawEvent {
    created_at: String,
    duration_seconds: Option<f64>,
    text_chars: Option<u32>,
}

#[derive(Debug, Clone)]
struct ParsedEvent {
    created_at: DateTime<Local>,
    duration_seconds: f64,
    text_chars: u32,
}

impl UsageStats {
    fn empty() -> Self {
        Self {
            session_count: 0,
            total_seconds: 0.0,
            total_chars: 0,
            total_minutes_int: 0,
            avg_chars_per_minute: 0.0,
        }
    }

    fn from_events<'a>(events: impl Iterator<Item = &'a ParsedEvent>) -> Self {
        let mut stats = Self::empty();
        for event in events {
            stats.session_count += 1;
            stats.total_seconds += event.duration_seconds.max(0.0);
            stats.total_chars += event.text_chars;
        }
        stats.total_minutes_int = (stats.total_seconds / 60.0).floor() as u32;
        stats.avg_chars_per_minute = if stats.total_seconds <= 0.0 {
            0.0
        } else {
            stats.total_chars as f64 / (stats.total_seconds / 60.0)
        };
        stats
    }
}

pub fn stats_path() -> PathBuf {
    let mut candidates = Vec::new();
    if let Ok(cwd) = std::env::current_dir() {
        candidates.push(cwd.join("voice_input_stats.jsonl"));
        candidates.push(cwd.join("..").join("voice_input_stats.jsonl"));
    }
    if let Ok(exe) = std::env::current_exe() {
        if let Some(dir) = exe.parent() {
            candidates.push(dir.join("voice_input_stats.jsonl"));
            candidates.push(
                dir.join("..")
                    .join("..")
                    .join("..")
                    .join("voice_input_stats.jsonl"),
            );
            candidates.push(
                dir.join("..")
                    .join("..")
                    .join("..")
                    .join("..")
                    .join("..")
                    .join("voice_input_stats.jsonl"),
            );
        }
    }

    for candidate in &candidates {
        if candidate.exists() {
            return dunce::simplified(candidate).to_path_buf();
        }
    }
    let fallback = candidates
        .first()
        .cloned()
        .unwrap_or_else(|| PathBuf::from("voice_input_stats.jsonl"));
    dunce::simplified(&fallback).to_path_buf()
}

pub fn load_stats_snapshot() -> StatsSnapshot {
    let path = stats_path();
    let events = read_events(&path);
    let now = Local::now();
    let recent_24h_start = now - Duration::hours(24);
    let recent_7d_start = now - Duration::days(7);
    let recent_24h = UsageStats::from_events(
        events
            .iter()
            .filter(|event| event.created_at >= recent_24h_start),
    );
    let recent_7d = UsageStats::from_events(
        events
            .iter()
            .filter(|event| event.created_at >= recent_7d_start),
    );

    let mut by_day = Vec::new();
    for offset in (0..7).rev() {
        let day = (now - Duration::days(offset)).date_naive();
        let day_key = day.to_string();
        let stats = UsageStats::from_events(
            events
                .iter()
                .filter(|event| event.created_at.date_naive() == day),
        );
        by_day.push(DailyUsageStats {
            day: day_key,
            stats,
        });
    }

    let history = events
        .iter()
        .rev()
        .take(30)
        .map(|event| HistoryEvent {
            created_at: event.created_at.format("%Y-%m-%d %H:%M:%S").to_string(),
            duration_seconds: event.duration_seconds,
            text_chars: event.text_chars,
        })
        .collect();

    StatsSnapshot {
        path: path.display().to_string(),
        recent_24h,
        recent_7d,
        by_day,
        history,
    }
}

pub fn append_event(text: &str, duration_seconds: f64) -> Result<(), String> {
    let path = stats_path();
    if let Some(parent) = path.parent() {
        std::fs::create_dir_all(parent).map_err(|err| format!("创建统计目录失败: {}", err))?;
    }
    let event = json!({
        "created_at": Local::now().format("%Y-%m-%dT%H:%M:%S").to_string(),
        "duration_seconds": (duration_seconds.max(0.0) * 100.0).round() / 100.0,
        "text_chars": count_text_chars(text),
    });
    let mut file = std::fs::OpenOptions::new()
        .create(true)
        .append(true)
        .open(&path)
        .map_err(|err| format!("打开统计文件失败: {}", err))?;
    writeln!(file, "{}", event).map_err(|err| format!("写入统计文件失败: {}", err))
}

fn count_text_chars(text: &str) -> u32 {
    text.split_whitespace().collect::<String>().chars().count() as u32
}

fn read_events(path: &PathBuf) -> Vec<ParsedEvent> {
    let Ok(text) = std::fs::read_to_string(path) else {
        return Vec::new();
    };

    text.lines()
        .filter_map(|line| serde_json::from_str::<RawEvent>(line).ok())
        .filter_map(|raw| {
            let created_at =
                chrono::NaiveDateTime::parse_from_str(&raw.created_at, "%Y-%m-%dT%H:%M:%S")
                    .ok()
                    .and_then(|value| value.and_local_timezone(Local).single())?;
            Some(ParsedEvent {
                created_at,
                duration_seconds: raw.duration_seconds.unwrap_or(0.0),
                text_chars: raw.text_chars.unwrap_or(0),
            })
        })
        .collect()
}
